### DISCUSSION TITLE
Getting 403 while accessing the image

### DISCUSSION THREAD

**Discussion by Zuhal-Ayob** [directus/directus#15726] [Archive]
### Describe the Bug

403 

{"errors":[{"message":"You don't have permission to access this.","extensions":{"code":"FORBIDDEN"}}]}

### To Reproduce

getting 403 while accessing the image , it happened after updating directus to the most recent version

### Errors Shown

{"errors":[{"message":"You don't have permission to access this.","extensions":{"code":"FORBIDDEN"}}]}

### What version of Directus are you using?

9.17.4

### What version of Node.js are you using?

16.17.0

### What database are you using?

Postgres 10

### What browser are you using?

Chrome

### How are you deploying Directus?

PaaS

**Comment by azrikahar** [comment]
There doesn't seem to be much info we can work with here, would you mind answering the following questions?

- What version were you updating _from_?
- By `PaaS` deployment, how are the assets stored? What is the storage location currently in use?
- Have you grant the necessary permissions to `directus_files`? What role are you accessing this image with?
- Can you share assets endpoint you have tried (eg. `/assets/c2e81a3f-4005-4ce3-8671-154d40ff197a`) or just the UUID of the image?

**Comment by br41nslug** [comment]
Closing this issue for now until more information is available

**Comment by Zuhal-Ayob** [comment]
> There doesn't seem to be much info we can work with here, would you mind answering the following questions?
> 
> * What version were you updating _from_?
> * By `PaaS` deployment, how are the assets stored? What is the storage location currently in use?
> * Have you grant the necessary permissions to `directus_files`? What role are you accessing this image with?
> * Can you share assets endpoint you have tried (eg. `/assets/c2e81a3f-4005-4ce3-8671-154d40ff197a`) or just the UUID of the image?

* I updated from `9.0.0-rc.101` , it was working with that version, after that, I haver tried several newer version which is not working.
*  Directus is hosted on Gov PaaS,  backed by Postgres, Redis, and S3. we are storing the videos on Vimeo (videos are accessible ) and the images in the Directus.
* yeah, it was working fine, it happened after updating the version of the Directus, I am accessing the images with the administrator user .
* This is the endpoint `/assets/8cc4f9b1-9867-498f-875c-8a0e5f55f401`

**Reply by azrikahar** [reply]
Unfortunately `9.0.0-rc.101` is quite a distance in terms of versions, so it could be hard to tell what could be the root cause here 🤔 

It _might_ be the same issue as #12474, based on the endpoint `https://sfccmstest.cloudapps.digital/assets/8cc4f9b1-9867-498f-875c-8a0e5f55f401` you shared. For some reason it's returning `content-type: application/json; charset=utf-8`, which is not an image. 

Aside from that, I also do notice there's a `x-vcap-request-id` response header, so if this is something to do with a reverse proxy somewhere in your setup, then you may need troubleshoot it to find out if it's the one causing issue here, similar to https://github.com/directus/directus/discussions/12474#discussioncomment-2466525.

**Reply by Zuhal-Ayob** [reply]
> Unfortunately `9.0.0-rc.101` is quite a distance in terms of versions, so it could be hard to tell what could be the root cause here 🤔
> 
> It _might_ be the same issue as #12474, based on the endpoint `https://sfccmstest.cloudapps.digital/assets/8cc4f9b1-9867-498f-875c-8a0e5f55f401` you shared. For some reason it's returning `content-type: application/json; charset=utf-8`, which is not an image.
> 
> Aside from that, I also do notice there's a `x-vcap-request-id` response header, so if this is something to do with a reverse proxy somewhere in your setup, then you may need troubleshoot it to find out if it's the one causing issue here, similar to [#12474 (comment)](https://github.com/directus/directus/discussions/12474#discussioncomment-2466525).

>* Thank you so much, at first the Direct us was updated to the `9.7.0 ` version which i had the same issue and decided to update it to the latest version which unfortunately didn't worked.

**Reply by Zuhal-Ayob** [reply]
The problem is it can't find the address of the image which is why it returns forbidden. if its a security issue why it was working before updating the version of DirectUs.

**Reply by azrikahar** [reply]
> The problem is it can't find the address of the image which is why it returns forbidden. 

This could very well be the issue, but there could still be (and has been reported) to be other issues so just trying not to rule them out 👍 We're discussing how to make it more verbose and not just the safe 403 error, particularly for admins, in #4368.

> if its a security issue why it was working before updating the version of DirectUs.

This is because Content Security Policy (CSP) headers were only added in a later version than `9.0.0-rc.101`. It's present in `9.4.2` onwards to be exact, added by this PR #10776.

---

Would you mind sharing your s3 storage related environment variables? Feel free to redact sensitive keys/secrets, just more so curious what's the current config in use.

**Reply by Zuhal-Ayob** [reply]
These are my s3 storage-related environment variables :
STORAGE_AMAZON_BUCKET,
STORAGE_AMAZON_DRIVER,
STORAGE_AMAZON_KEY,
STORAGE_AMAZON_REGION,
STORAGE_AMAZON_SECRET,
STORAGE_LOCATIONS

**Reply by azrikahar** [reply]
Can you share the values of said environment variables? You can redact the key, secret and domain name (if applicable).

**Reply by Zuhal-Ayob** [reply]
Here i am attaching the values of the environment variables :
STORAGE_AMAZON_BUCKET: paas-s3-broker-prod-98c3d8ee-2f84-423c-a3f9-0501662836dc
STORAGE_AMAZON_DRIVER: s3
STORAGE_AMAZON_REGION: eu-west-1
STORAGE_LOCATIONS: amazon

**Reply by azrikahar** [reply]
I'm not noticing an obvious issue with the environment variables, but would you be able to try something like aws-cli to check whether it's a permission/ACL issue for that s3 bucket?

**Reply by Zuhal-Ayob** [reply]
I don't see any problem with accessing s3 ,  I think the problem is with Content Security Policy (CSP) headers which is recently introduced and  i can't set it to default because i don't have any header set up to change the response header and as you mentioned the solution on this ticket #12474 unfortunately i dont have any nginx configuration too.

**Reply by azrikahar** [reply]
Can you try setting `ASSET_CONTENT_SECURITY_POLICY` environment variables (depending on the CSP error you are facing)? This linked comment elaborates more on how to match it with Helmet.js configurations: https://github.com/directus/directus/discussions/15865#discussioncomment-3786705

**Reply by Zuhal-Ayob** [reply]
Setting the environment variables changed the CSP for me, but still, the images aren’t loading for me , not sure about the value of CSP, just copied them from helmetjs.

These are the environment variables:

ASSETS_CONTENT_SECURITY_POLICY_DIRECTIVES__DEFAULT_SRC: default-src 'self'
ASSETS_CONTENT_SECURITY_POLICY_DIRECTIVES__OBJECT_SRC: object-src 'none'
ASSETS_CONTENT_SECURITY_POLICY_DIRECTIVES__SCRIPT_SRC: script-src 'self'
ASSETS_CONTENT_SECURITY_POLICY_DIRECTIVES__UPGRADE_INSECURE_REQUESTS: upgrade-insecure-requests
ASSETS_CONTENT_SECURITY_POLICY_USE_DEFAULT: false


<img width="1422" alt="Screenshot 2022-10-18 at 23 56 49" src="https://user-images.githubusercontent.com/90700942/196560501-f5c599e5-2bcd-4477-b9d1-9854901c3b4d.png">

**Reply by azrikahar** [reply]
You can view the browser devtools to verify whether the CSP are correct 👍 In this case, you are duplicating the directives (for example, `default-src` shouldn't be included for `ASSETS_CONTENT_SECURITY_POLICY_DIRECTIVES__DEFAULT_SRC`) so they are all currently malformed:

![](https://user-images.githubusercontent.com/42867097/196600643-c5ebbf5a-054c-4f57-b1ef-90778313cf2e.png)

you can try removing the ones you are currently using, and based the errors I'm seeing, you can test these environment variables first:

```yaml
ASSETS_CONTENT_SECURITY_POLICY_DIRECTIVES__DEFAULT_SRC=array:'self',data:,blob:,https://cdn.directus.io
ASSETS_CONTENT_SECURITY_POLICY_DIRECTIVES__STYLE_SRC=array:'self','unsafe-inline'
```

then do make sure to modify/add other directives depending on what other CSP error comes up in your browser devtools afterwards.

**Reply by Zuhal-Ayob** [reply]
Thank you so much, it worked, and the images are loading for me now.

**Comment by LukasV** [comment]
That's right. Editing .env file for directus as follows fixed the issue for me:

`CONTENT_SECURITY_POLICY_DIRECTIVES__IMG_SRC="array:'self',youtube.com,ytimg.com,vimeo.com,mydomain.com,*.youtube.com,*.ytimg.com,*.vimeo.com,*.mydomain.com,data:,blob:,https://cdn.directus.io"
CONTENT_SECURITY_POLICY_DIRECTIVES__CHILD_SRC="array:'self',youtube.com,ytimg.com,vimeo.com,mydomain.com,*.youtube.com,*.ytimg.com,*.vimeo.com,*.mydomain.com"
CONTENT_SECURITY_POLICY_DIRECTIVES__FRAME_ANCESTORS="array:'self',youtube.com,ytimg.com,vimeo.com,mydomain.com,*.youtube.com,*.ytimg.com,*.vimeo.com,*.mydomain.com"`
